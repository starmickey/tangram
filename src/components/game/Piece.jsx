import '../../styles/pieceType.css'

export default ({ type, left, top }) => {
    let {src, className} = type;
    let style = {left: left, top: top}

    return (
        <div>
            <img
                src={src}
                className={className}
                style = {style}
                alt="" />
        </div>
    )

}

