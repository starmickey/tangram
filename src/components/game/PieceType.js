export default {
    STRIANGLE: {
        id: 0,
        className: "striangle",
        src: "./striangle.png"
    },
    MTRIANGLE: {
        id: 1,
        className: "mtriangle",
        src: "./mtriangle.png"
    },
    LTRIANGLE: {
        id: 2,
        className: "ltriangle",
        src: "./ltriangle.png"
    },
    PARALLELOGRAM: {
        id: 3,
        className: "parallelogram",
        src: "./parallelogram.png"
    },
    SQUARE: {
        id: 4,
        className: "square",
        src: "./square.png"
    },
}