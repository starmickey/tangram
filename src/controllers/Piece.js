export default class Piece {
  constructor(id, typeId, x = 0, y = 0, a = 0) {
    this.id = id;
    this.typeId = typeId;
    this.x = x;
    this.y = y;
    this.a = a;
  }

  setPosition(x, y) {
    this.x = x;
    this.y = y;
    return this;
  }

  setA(a) {
    this.a = a;
    return this;
  }
}
