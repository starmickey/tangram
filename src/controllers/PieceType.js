const PieceType = {
  STRIANGLE: {
    id: 0,
    src: "./triangle.png",
    height: 70.71,
    width: 141.42,
  },
  MTRIANGLE: {
    id: 1,
    src: "./triangle.png",
    height: 100,
    width: 200,
  },
  LTRIANGLE: {
    id: 2,
    src: "./triangle.png",
    height: 141.42,
    width: 282.84,
  },
  PARALLELOGRAM: {
    id: 3,
    src: "./parallelogram.png",
    height: 100,
    width: 200,
  },
  SQUARE: {
    id: 4,
    src: "./square.png",
    height: 100,
    width: 100,
  },
};

export const getTypeById = (id) => {
  switch (id) {
    case 0:
      return PieceType.STRIANGLE;
    case 1:
      return PieceType.MTRIANGLE;
    case 2:
      return PieceType.LTRIANGLE;
    case 3:
      return PieceType.PARALLELOGRAM;
    case 4:
      return PieceType.SQUARE;
    default:
      return -1;
  }
};

export default PieceType;
