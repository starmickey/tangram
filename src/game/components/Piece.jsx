import PropTypes from "prop-types";
import { useState } from "react";
import GameHandler from "../controllers/GameHandler";
import useDragAndClick from "./utils/useDragAndClick";
import PieceComponent from "./utils/PieceComponent";

/**
 * Renders a dynamic piece component
 * @param {number} pieceId - unique piece id
 * @param {number} gameHandler - its passed because if the
 * piece state changes, it must be updated on the gameHandler
 * @param {func} handleGameChange - changes the parent compoment state
 */

function Piece({ pieceId, gameHandler, handleGameChange }) {
  // Validate inputs
  if (pieceId < 0) {
    throw new Error("Invalid input parameters");
  }

  // Get actual piece state from Game Handler
  const pieceDTO = gameHandler.getPieceDTO(pieceId);
  // Create piece status hook
  const [piece, setPiece] = useState(pieceDTO);

  // Get events handlers
  const {
    isDragging,
    handleDragStart,
    handleDragEnd,
    handleClick,
  } = useDragAndClick(pieceId, setPiece, gameHandler, handleGameChange);

  // Create handlers package for creating piececomponent
  const handlers = {
    handleDragStart,
    handleDragEnd,
    handleClick,
  };

  // Custom pieces styles
  const styles = {
    fill: "#00D2FF",
    stroke: ":#000000",
    strokeWidth: 0.1,
    shadowOffset: 0.5,
    shadowOffsetOnDrag: 1,
    shadowBlur: 3,
    shadowBlurOnDrag: 5,
    scaleOnDrag: 1.05,
  };

  // Render piece component
  return (
    <PieceComponent
      piece={piece}
      isDragging={isDragging}
      handlers={handlers}
      styles={styles}
    />
  );
}

Piece.propTypes = {
  pieceId: PropTypes.number.isRequired,
  gameHandler: PropTypes.instanceOf(GameHandler).isRequired,
  handleGameChange: PropTypes.func.isRequired,
};

export default Piece;
