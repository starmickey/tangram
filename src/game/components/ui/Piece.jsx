import PropTypes from "prop-types";
import { useState, useRef } from "react";
import { Shape } from "react-konva";
import getCorners from "../utils/getCornersStrategy";
import { useScaleState } from "../../contexts/StageContext";

/**
 * React component that displace a "piece",
 * which is a geometric figure
 */
function Piece({
  typeId,
  width,
  height,
  x,
  y,
  a,
  fill,
  stroke,
  strokeWidth,
  shadowOffset,
  shadowBlur,
  scale,
  draggable = false,
  handleDragStart,
  handleDragEnd,
  handleDragBound,
  handleClick,
}) {
  // Position hook. We need it because the piece may be draggable
  // Also, events may update the position.
  const [position, setPosition] = useState({ x, y, a });
  // Piece reference for accesing attributes after dragging
  const pieceRef = useRef();

  // Get context for piece styling
  const contextScale = useScaleState();

  // Auxiliary functions for piece handling
  const getPosition = () => ({
    x: pieceRef.current.x(),
    y: pieceRef.current.y(),
    a: pieceRef.current.rotation(),
  });

  return (
    <Shape
      // Draw the piece
      sceneFunc={(context, shape) => {
        const corners = getCorners(
          typeId,
          width,
          height,
        );
        const lastCorner = corners[corners.length - 1];
        context.beginPath();
        context.moveTo(lastCorner.x, lastCorner.y);
        corners.forEach((c) => {
          const { x: cx, y: cy } = c.getPosition();
          context.lineTo(cx, cy);
        });
        context.closePath();
        context.fillStrokeShape(shape);
      }}
      // Position
      x={position.x}
      y={position.y}
      rotation={position.a}
      // Dimensions
      width={width}
      height={height}
      // Format
      fill={fill}
      stroke={stroke}
      strokeWidth={strokeWidth}
      // Ensure it rotates around its center
      offsetX={width / 2}
      offsetY={height / 2}
      // Shadow
      shadowOffsetX={shadowOffset}
      shadowOffsetY={shadowOffset}
      shadowBlur={shadowBlur}
      // Scale
      scaleX={contextScale * scale}
      scaleY={contextScale * scale}
      // Set if its draggable
      draggable={draggable}
      // Event handlers
      onClick={
        draggable
          ? () => {
            const newPos = getPosition();
            setPosition(newPos);
            handleClick(newPos, setPosition);
          }
          : null
      }
      dragBoundFunc={
        draggable
          ? (pos) => {
            // Get the piece dimensions
            const { width: pieceWidth, height: pieceHeight } = pieceRef.current.getClientRect();
            // Get position after handling dragging bounds
            const dim = { width: pieceWidth, height: pieceHeight };
            const newPos = handleDragBound(pos, dim);
            return newPos;
          }
          : null
      }
      onDragStart={
        draggable
          ? () => handleDragStart(position, setPosition)
          : null
      }
      onDragEnd={
        draggable
          ? () => {
            const newPos = getPosition();
            setPosition(newPos);
            handleDragEnd(newPos, setPosition);
          }
          : null
      }
      ref={pieceRef}
    />
  );
}

Piece.defaultProps = {
  width: 100,
  height: 100,
  x: 0,
  y: 0,
  a: 0,
  fill: "#000000",
  stroke: "#000000",
  strokeWidth: 0.1,
  shadowOffset: 0,
  shadowBlur: 0,
  scale: 1,
  draggable: false,
  handleDragStart: () => { },
  handleDragEnd: () => { },
  handleDragBound: () => { },
  handleClick: () => { },
};

Piece.propTypes = {
  typeId: PropTypes.number.isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  x: PropTypes.number,
  y: PropTypes.number,
  a: PropTypes.number,
  fill: PropTypes.string,
  stroke: PropTypes.string,
  strokeWidth: PropTypes.number,
  shadowOffset: PropTypes.number,
  shadowBlur: PropTypes.number,
  scale: PropTypes.number,
  draggable: PropTypes.bool,
  handleDragStart: PropTypes.func,
  handleDragEnd: PropTypes.func,
  handleDragBound: PropTypes.func,
  handleClick: PropTypes.func,
};

export default Piece;
