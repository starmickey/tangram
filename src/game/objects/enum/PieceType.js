const PieceType = {
  STRIANGLE: {
    id: 0,
    height: 70.71,
    width: 141.42,
  },
  MTRIANGLE: {
    id: 1,
    height: 100,
    width: 200,
  },
  LTRIANGLE: {
    id: 2,
    height: 141.42,
    width: 282.84,
  },
  PARALLELOGRAM: {
    id: 3,
    height: 70.71,
    width: 212.13,
  },
  SQUARE: {
    id: 4,
    height: 100,
    width: 100,
  },
};

export function getPieceType(id) {
  switch (id) {
    case PieceType.STRIANGLE.id:
      return PieceType.STRIANGLE;
    case PieceType.MTRIANGLE.id:
      return PieceType.MTRIANGLE;
    case PieceType.LTRIANGLE.id:
      return PieceType.LTRIANGLE;
    case PieceType.PARALLELOGRAM.id:
      return PieceType.PARALLELOGRAM;
    case PieceType.SQUARE.id:
      return PieceType.SQUARE;
    default:
      throw new Error(`cannot get a piece type for id ${id}`);
  }
}

export default PieceType;
