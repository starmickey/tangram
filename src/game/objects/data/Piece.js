import PieceType from "../enum/PieceType";

export default class Piece {
  static nextId = 0;

  /**
   * Create a new Piece object
   * @param {PieceType} type - the piece type
   * @param {number} x - X-coordinate
   * @param {number} y - Y-coordinate
   * @param {number} a - rotation
   */
  constructor(type, x = 0, y = 0, a = 0) {
    // Validate inputs
    if (
      !(type instanceof PieceType)
      || typeof x !== "number"
      || typeof y !== "number"
      || typeof a !== "number"
      || a < 0
      || a > 360
    ) {
      throw new Error("Invalid input parameters");
    }

    // Assign attributes
    this.id = Piece.#getNextId();
    this.type = type;
    this.x = x;
    this.y = y;
    this.a = a;
  }

  // Generate unique ids
  static #getNextId() {
    this.nextId += 1;
    return this.nextId;
  }

  /**
   * Update piece position
   * @param {number} x - new X-coordinate
   * @param {number} y - new Y-coordinate
   */
  setPosition(x, y) {
    // Validate inputs
    if (
      typeof x !== "number"
      || typeof y !== "number"
    ) {
      throw new Error("Invalid inputs");
    }

    // Assign attributes
    this.x = x;
    this.y = y;
  }

  /**
   * Rotate piece
   * @param {number} a - new angle
   */
  setA(a) {
    // Validate inputs
    if (
      typeof a !== "number"
      || a < 0
      || a > 360
    ) {
      throw new Error(`Invalid angle ${a}`);
    }

    // Assign attributes
    this.a = a;
  }
}
