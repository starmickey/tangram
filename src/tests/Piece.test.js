import renderer from 'react-test-renderer';
import Piece from '../components/game/Piece';
import PieceType from '../components/game/PieceType';


it('renders Piece', () => {
    const piece = renderer.create(
        <Piece
            type={PieceType.STRIANGLE}
            left="20rem"
            right="32rem"
         />
    );
    let tree = piece.toJSON();
    expect(tree).toMatchSnapshot();
})
