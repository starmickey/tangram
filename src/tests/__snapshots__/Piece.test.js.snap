// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Page component 1`] = `
<div>
  <img
    alt=""
    className="striangle"
    src="./striangle.png"
  />
</div>
`;
